pinot-core/src/main/java/org/apache/pinot/core/realtime/impl/dictionary/OffHeapMutableBytesStore.java

```diff
@@ -0,0 +1,195 @@
+    int offsetIndex = _numValues & OFFSET_BUFFER_MASK;
+
+    PinotDataBuffer offsetBuffer;
+    // If this is the first entry in the offset buffer, allocate a new buffer and store the start offset (end offset of
+    // the previous value)
+    if (offsetIndex == 0) {
+      offsetBuffer = _memoryManager.allocate(OFFSET_BUFFER_SIZE, _allocationContext);
+      offsetBuffer.putInt(0, _valueStartOffset);
+      _offsetBuffers[offsetBufferIndex] = offsetBuffer;
+      _totalBufferSize += OFFSET_BUFFER_SIZE;
+    } else {
+      offsetBuffer = _offsetBuffers[offsetBufferIndex];
+    }
+
+    int valueLength = value.length;
+    if (valueLength == 0) {
+      offsetBuffer.putInt((offsetIndex + 1) << 2, _valueStartOffset);
+      return _numValues++;
+    }
+
+    int valueBufferIndex = (_valueStartOffset + valueLength - 1) >>> VALUE_BUFFER_SIFT_OFFSET;
+    PinotDataBuffer valueBuffer;
+    // If the current value buffer does not have enough space, allocate a new buffer to store the value
+    if ((_valueStartOffset - 1) >>> VALUE_BUFFER_SIFT_OFFSET != valueBufferIndex) {
+      valueBuffer = _memoryManager.allocate(VALUE_BUFFER_SIZE, _allocationContext);
+      _valueBuffers[valueBufferIndex] = valueBuffer;
+      _totalBufferSize += VALUE_BUFFER_SIZE;
+      _valueStartOffset = valueBufferIndex << VALUE_BUFFER_SIFT_OFFSET;
+    } else {
+      valueBuffer = _valueBuffers[valueBufferIndex];
+    }
+
+    int valueEndOffset = _valueStartOffset + valueLength;
+    offsetBuffer.putInt((offsetIndex + 1) << 2, valueEndOffset);
+    valueBuffer.readFrom(_valueStartOffset & VALUE_BUFFER_MASK, value);
+    _valueStartOffset = valueEndOffset;
+
+    return _numValues++;
+  }
+
+  public byte[] get(int index) {
+    int offsetBufferIndex = index >>> OFFSET_BUFFER_SHIFT_OFFSET;
+    int offsetIndex = index & OFFSET_BUFFER_MASK;
+    int previousValueEndOffset = _offsetBuffers[offsetBufferIndex].getInt(offsetIndex << 2);
+    int valueEndOffset = _offsetBuffers[offsetBufferIndex].getInt((offsetIndex + 1) << 2);
+
+    if (previousValueEndOffset == valueEndOffset) {
+      return EMPTY_BYTES;
+    }
+
+    int valueBufferIndex = (valueEndOffset - 1) >>> VALUE_BUFFER_SIFT_OFFSET;
+    int startOffsetInValueBuffer;
+    int valueLength;
+    if ((previousValueEndOffset - 1) >>> VALUE_BUFFER_SIFT_OFFSET != valueBufferIndex) {
+      // The first value in the value buffer
+      startOffsetInValueBuffer = 0;
+      valueLength = valueEndOffset & VALUE_BUFFER_MASK;
+    } else {
+      // Not the first value in the value buffer
+      startOffsetInValueBuffer = previousValueEndOffset & VALUE_BUFFER_MASK;
+      valueLength = valueEndOffset - previousValueEndOffset;
+    }
+
+    byte[] value = new byte[valueLength];
+    _valueBuffers[valueBufferIndex].copyTo(startOffsetInValueBuffer, value);
+    return value;
+  }
+
+  public boolean equals(int index, byte[] value) {
+    int valueLength = value.length;
```

Comment:
final. Also, it may be useful to rename this as inputValueLength (and 'value' as 'inputValue') fpor readability

Reply:
I'm not a fan of using final on local variables. Changed the valueLength to inputValueLength.